tests/integration/test_healing.py

```diff
@@ -0,0 +1,281 @@
+    if app := await app_name(ops_test):
+        if len(ops_test.model.applications[app].units) == 3:
+            return
+        else:
+            async with ops_test.fast_forward():
+                await scale_application(ops_test, app, 3)
+            return
+
+    # Build and deploy charm from local source folder
+    charm = await ops_test.build_charm(".")
+
+    # Reduce the update_status frequency until the cluster is deployed
+    async with ops_test.fast_forward():
+        await ops_test.model.deploy(charm, application_name=APP_NAME, num_units=3)
+
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[APP_NAME].units) == 3
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+
+
+@pytest.mark.order(1)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_kill_db_process(ops_test: OpsTest) -> None:
+    """Kill mysqld process and check for auto cluster recovery."""
+    await build_and_deploy(ops_test)
+
+    app = await app_name(ops_test)
+    unit = ops_test.model.applications[app].units[0]
+    another_unit = ops_test.model.applications[app].units[1]
+
+    # get running mysqld PID
+    pid = await get_process_pid(ops_test, unit.name, MYSQL_DAEMON)
+
+    # kill mysqld for the unit
+    logger.info(f"Killing process id {pid}")
+    await ops_test.juju("ssh", unit.name, "sudo", "kill", "-9", pid)
+
+    config = {
+        "username": CLUSTER_ADMIN_USERNAME,
+        "password": await get_system_user_password(unit, CLUSTER_ADMIN_USERNAME),
+        "host": await unit.get_public_address(),
+    }
+
+    # verify connection so daemon is restarted
+    assert is_connection_possible(config), "❌ Daemon did not restarted"
+
+    # retrieve new PID
+    new_pid = await get_process_pid(ops_test, unit.name, MYSQL_DAEMON)
+    logger.info(f"New process id is {new_pid}")
+
+    # verify that mysqld instance is not the killed one
+    assert new_pid != pid, "❌ PID for mysql daemon did not changed"
+
+    # verify instance is part of the cluster
+    logger.info("Check if instance back in cluster")
+    assert await is_unit_in_cluster(
+        ops_test, unit.name, another_unit.name
+    ), " Unit not online in the cluster"
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_freeze_db_process(ops_test: OpsTest):
+    """Freeze and unfreeze process and check for auto cluster recovery."""
+    app = await app_name(ops_test)
+    unit = ops_test.model.applications[app].units[0]
+    another_unit = ops_test.model.applications[app].units[1]
+
+    # get running mysqld PID
+    pid = await get_process_pid(ops_test, unit.name, MYSQL_DAEMON)
+
+    # freeze (STOP signal) mysqld for the unit
+    logger.info(f"Freezing process id {pid}")
+    await ops_test.juju("ssh", unit.name, "sudo", "kill", "-19", pid)
```

Comment:
see above

Reply:
Kind the same, I validate the stop when asserting connection is not possible (L133)